examples/pna-example-tcp-connection-state-tracking-simplified.p4

```diff
@@ -0,0 +1,249 @@
+
+    // Outputs from actions of ct_tcp_table
+    AddEntryErrorStatus_t add_status;
+
+
+    action ct_tcp_table_hit (
+        inout bit<32> n2h_seqNo; // NEW: support for write back entry
+        inout bit<32> h2n_seqNo; // NEW: support for write back entry
+        inout bit<32> n2h_ackNo; // NEW: support for write back entry
+        inout bit<32> h2n_ackNo; // NEW: support for write back entry
+        // More action data that is not written back can be here as well
+        ) {
+        // some table types, e.g., T-CAM-based ones, may not support re-writable
+        // entries.
+        if ((hdr.tcp.flags & TCP_SYN_MASK) != 0) {
+          // the code for handling SYN messages for sessions already in the TCP
+          // session table goes here
+        } else {
+            if (hdr.tcp.ackNo<=SelectByDirection(istd.direction,h2n_seqNo,n2h_seqNo) &
+                hdr.tcp.ackNo>=SelectByDirection(istd.direction,n2h_ackNo,h2n_ackNo)) {
+                if (istd.direction==PNA_Direction_t.NET_TO_HOST) {
+                    n2h_seqNo=hdr.tcp.seqNo; // NEW: support for write back entry
+                    n2h_ackNo=hdr.tcp.ackNo; // NEW: support for write back entry
+// Mario to check that in 3-way handshake ackNo of forward direction is expected to be the last sequence number
+                } else {
+                    h2n_seqNo=hdr.tcp.seqNo; // NEW: support for write back entry
+                    h2n_ackNo=hdr.tcp.ackNo; // NEW: support for write back entry
+// Mario to check that in 3-way handshake ackNo of forward direction is expected to be the last sequence number
+                }
+                set_entry_expire_time(EXPIRE_TIME_PROFILE_TCP_ESTABLISHED;
+              } else
+                  drop_packet();
+        }
+    }
+
+    action ct_tcp_table_miss() {
+      // the code for creating entries in the TCP
+      // session table goes here
+    }
+
+    table ct_tcp_table {
+        /* add_on_miss table is restricted to have all exact match fields */
+        key = {
+            // other key fields also possible, e.g. VRF
+            SelectByDirection(istd.direction, hdr.ipv4.srcAddr, hdr.ipv4.dstAddr):
+                exact @name("ipv4_addr_0");
+            SelectByDirection(istd.direction, hdr.ipv4.dstAddr, hdr.ipv4.srcAddr):
+                exact @name("ipv4_addr_1");
+            hdr.ipv4.protocol : exact;
+            SelectByDirection(istd.direction, hdr.tcp.srcPort, hdr.tcp.dstPort):
+                exact @name("tcp_port_0");
+            SelectByDirection(istd.direction, hdr.tcp.dstPort, hdr.tcp.srcPort):
+                exact @name("tcp_port_1");
+        }
+        actions = {
+            @tableonly   ct_tcp_table_hit;
+            @defaultonly ct_tcp_table_miss;
+        }
+
+        add_on_miss = true;
+
+        default_idle_timeout_for_data_plane_added_entries = 1;
+
+        idle_timeout_with_auto_delete = true;
+        const default_action = ct_tcp_table_miss;
```

Comment:
how do we specify 'lock' property on the table so that any concurrent access to the table are serialized to ensure data integrity? Can we add a property on the table? Or is it expected to be implicitly handled by compiler? What if programmer does not require locking?

Reply:
I think a fairly straightforward way to do this is to define this feature so that the program must behave AS IF there were a per-table-entry lock that is acquired when the action begins executing, and is released when the action is finished executing.

Implementers might find other actual implementation methods that achieve the same behavior, and the spec should not restrict those, as long as the behavior is as described above.